src/group_theory/subgroup.lean

```diff
@@ -1662,6 +1717,143 @@ set.subset.antisymm
+  simp only [inv_mul_cancel_right],
+end
 
 end pointwise
+
+section subgroup_normal
+
+@[to_additive] lemma normal_of_iff {H K : subgroup G} (hHK : H ≤ K) :
+  (H.subgroup_of K).normal ↔ ∀ h k, h ∈ H → k ∈ K → k * h * k⁻¹ ∈ H :=
+⟨λ hN h k hH hK, hN.conj_mem ⟨h, hHK hH⟩ hH ⟨k, hK⟩,
+  λ hN, { conj_mem := λ h hm k, (hN h.1 k.1 hm k.2) }⟩
+
+@[to_additive] lemma prod_of_prod_normal
+  {H₁ K₁ : subgroup G} {H₂ K₂ : subgroup N}
+  [h₁ : (H₁.subgroup_of K₁).normal] [h₂ : (H₂.subgroup_of K₂).normal] :
+  ((H₁.prod H₂).subgroup_of (K₁.prod K₂)).normal :=
+{ conj_mem := λ n hgHK g,
+    ⟨h₁.conj_mem ⟨(n : G × N).fst, (mem_prod.mp n.2).1⟩
+    hgHK.1 ⟨(g : G × N).fst, (mem_prod.mp g.2).1⟩,
+    h₂.conj_mem ⟨(n : G × N).snd, (mem_prod.mp n.2).2⟩
+    hgHK.2 ⟨(g : G × N).snd, (mem_prod.mp g.2).2⟩⟩ }
+
+@[to_additive] instance prod_normal
+  (H : subgroup G) (K : subgroup N) [hH : H.normal] [hK : K.normal] :
+  (H.prod K).normal :=
+{ conj_mem := begin intros n hg g, rw subgroup.mem_prod,
+    exact ⟨hH.conj_mem n.fst (subgroup.mem_prod.mp hg).1 g.fst,
+      hK.conj_mem n.snd (subgroup.mem_prod.mp hg).2 g.snd⟩ end }
+
+@[to_additive] lemma inf_normal_inf_right
+  (A B' B : subgroup G) (hB : B' ≤ B) [hN : (B'.subgroup_of B).normal] :
+  ((A ⊓ B').subgroup_of (A ⊓ B)).normal :=
+{ conj_mem := λ n hn g,
+   ⟨mul_mem A (mul_mem A (mem_inf.1 g.2).1 (mem_inf.1 n.2).1) (inv_mem A (mem_inf.1 g.2).1),
+   (normal_of_iff hB).mp hN n g hn.2 (mem_inf.mp g.2).2⟩ }
+
+@[to_additive] lemma inf_normal_inf_left
+  {A' A : subgroup G} (B : subgroup G) (hA : A' ≤ A) [hN : (A'.subgroup_of A).normal] :
+  ((A' ⊓ B).subgroup_of (A ⊓ B)).normal :=
+{ conj_mem := λ n hn g,
+    ⟨(normal_of_iff hA).mp hN n g hn.1  (mem_inf.mp g.2).1,
+    mul_mem B (mul_mem B (mem_inf.1 g.2).2 (mem_inf.1 n.2).2) (inv_mem B (mem_inf.1 g.2).2)⟩ }
+
+instance normal_sup_normal (H K : subgroup G) [hH : H.normal] [hK : K.normal] : (H ⊔ K).normal :=
+{ conj_mem := λ n hmem g,
+  begin
+    change n ∈ ↑(H ⊔ K) at hmem,
+    change g * n * g⁻¹ ∈ ↑(H ⊔ K),
+    rw normal_mul at *,
+    rw set.mem_mul at *,
+    rcases hmem with ⟨h, k, hh, hk, hhk⟩,
+    refine ⟨g * h * g⁻¹, g * k * g⁻¹, hH.conj_mem h hh g, hK.conj_mem k hk g, _⟩,
+    rw ← hhk,
+    simp,
+  end }
+
+@[to_additive] instance normal_inf_normal (H K : subgroup G) [hH : H.normal] [hK : K.normal] :
+  (H ⊓ K).normal :=
+{ conj_mem := λ n hmem g,
+  by { rw mem_inf at *, exact ⟨hH.conj_mem n hmem.1 g, hK.conj_mem n hmem.2 g⟩ } }
+
+@[to_additive] lemma sup_of (A A' B : subgroup G) (hA : A ≤ B) (hA' : A' ≤ B) :
+  (A ⊔ A').subgroup_of B = A.subgroup_of B ⊔ A'.subgroup_of B :=
+begin
+  refine eq_of_map_eq_le_ker B.subtype
+    (comap B.subtype (A ⊔ A')) (comap B.subtype A ⊔ comap B.subtype A')
+    _ (ker_le_comap _ _) (le_trans (ker_le_comap B.subtype _) le_sup_left),
+  { simp only [subgroup_of, map_comap_eq, map_sup, monoid_hom.subtype_range],
+    rw [inf_of_le_right (sup_le hA hA'), inf_of_le_right hA', inf_of_le_right hA] },
```

Comment:
Here's an example of where we want `eq_of_map_eq_le_ker` to have less explicit parameters:
```suggestion
  refine eq_of_map_eq_le_ker _ _ _ _
    (ker_le_comap _ _) (le_trans (ker_le_comap B.subtype _) le_sup_left),
  { simp only [subgroup_of, map_comap_eq, map_sup, monoid_hom.subtype_range],
    rw [inf_of_le_right (sup_le hA hA'), inf_of_le_right hA', inf_of_le_right hA] },
```